### PR TITLE
impl(common): helpers to create ErrorInfo

### PR DESCRIPTION
**Pull request by coryan**
Motivated by #5915

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271)

### CODE REVIEW

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/internal/make_status_test.cc` line 135 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271#-NHB69UdEaWXxPbQRxGz:-NHB69UdEaWXxPbQRxH-:b-ddt0nl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6e94f9ab7875b5b1bd0449de168b97c15b39d79a/google/cloud/internal/make_status_test.cc#L135)):*
> ```cpp
> }
> 
> TEST(MakeStatus, WithErrorInfo) {
> ```

I'm curious why not use gTest's INSTANTIATE_TEST_P to do these parameterized tests?

**Review by scotthart — APPROVED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

**Review by coryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`google/cloud/internal/make_status_test.cc` line 135 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10271#-NHB69UdEaWXxPbQRxGz:-NHB78eG3eBQ6yGZ7Xj7:bo8kv0j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6e94f9ab7875b5b1bd0449de168b97c15b39d79a/google/cloud/internal/make_status_test.cc#L135)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I'm curious why not use gTest's INSTANTIATE_TEST_P to do these parameterized tests?
</blockquote></details>

I keep that at the bottom of the toolbox. For not particular reason.

File: google/cloud/internal/make_status.h [hunk 1]
```diff
@@ -41,6 +42,87 @@ Status InternalError(std::string msg, ErrorInfo info = {});
...
+ *   return i->get<std::string>(); // no type checking in this example
+ * }
+ * @endcode
+ *
+ */
+class ErrorInfoBuilder {
+ public:
+  ErrorInfoBuilder(std::string file, int line, std::string function);
+
+  /// Adds the metadata from an error context, existing values are not replaced.
```
Comment (devbww) [inline google/cloud/internal/make_status.h:81]: Perhaps `metadata_` should be an (unordered) multimap??

Comment (devbww) [reply google/cloud/internal/make_status.h:81]: Ack.

**Pull request merged**